test_fitting_with_constrained_warp appears to be an unflagged flaky

### What's the problem?

`tests.units.warps.test_cwgp_training.ConstraintTests.test_fitting_with_constrained_warp` failed when testing from requirements file - see https://github.com/gchq/Vanguard/actions/runs/8603220003/attempts/1?pr=77. It passed on rerun. This test is not marked as flaky.

### How can we reproduce the issue?

https://github.com/gchq/Vanguard/actions/runs/8603220003/attempts/1?pr=77

### Python version

3.9

### Package version

_No response_

### Operating system

Linux - GitHub runner

### Other packages

_No response_

### Relevant log output

______________ ConstraintTests.test_fitting_with_constrained_warp ______________

self = <tests.units.warps.test_cwgp_training.ConstraintTests testMethod=test_fitting_with_constrained_warp>

    def test_fitting_with_constrained_warp(self):
        """Should NOT throw a RuntimeError."""
        box_cox = warpfunctions.BoxCoxWarpFunction(lambda_=0)
        affine = warpfunctions.PositiveAffineWarpFunction(a=1, b=2)
    
        @SetWarp(box_cox @ affine, ignore_methods=("__init__",))
        class TestController(GaussianGPController):
            """A test controller."""
            pass
    
        scaler = StandardScaler()
        gp = TestController(scaler.fit_transform(self.DATASET.train_x), self.DATASET.train_y, ScaledRBFKernel,
                            y_std=self.DATASET.train_y_std)
    
        try:
>           gp.fit(100)

tests/units/warps/test_cwgp_training.py:297: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
vanguard/base/gpcontroller.py:129: in fit
    loss = self._sgd_round(n_iters=n_sgd_iters, gradient_every=gradient_every)
vanguard/warps/decorator.py:122: in _sgd_round
    loss = super()._sgd_round(n_iters=n_iters, gradient_every=gradient_every)
vanguard/base/basecontroller.py:323: in _sgd_round
    raise err
vanguard/base/basecontroller.py:308: in _sgd_round
    loss = self._single_optimisation_step(train_x, train_y, retain_graph=(iter_num < n_iters - 1))
vanguard/base/basecontroller.py:347: in _single_optimisation_step
    loss = self._loss(x, y)
vanguard/warps/decorator.py:[139](https://github.com/gchq/Vanguard/actions/runs/8603220003/job/23574608334#step:6:140): in _loss
    nmll = super()._loss(train_x, warped_train_y)
vanguard/base/basecontroller.py:368: in _loss
    return -self._mll(output, train_y.squeeze(dim=-1))
vanguard/base/basecontroller.py:530: in __call__
    result = super().__call__(*args, **kwargs)
../../_tool/Python/3.9.18/x64/lib/python3.9/site-packages/gpytorch/module.py:30: in __call__
    outputs = self.forward(*inputs, **kwargs)
../../_tool/Python/3.9.18/x64/lib/python3.9/site-packages/gpytorch/mlls/exact_marginal_log_likelihood.py:62: in forward
    res = output.log_prob(target)
../../_tool/Python/3.9.18/x64/lib/python3.9/site-packages/gpytorch/distributions/multivariate_normal.py:169: in log_prob
    inv_quad, logdet = covar.inv_quad_logdet(inv_quad_rhs=diff.unsqueeze(-1), logdet=True)
../../_tool/Python/3.9.18/x64/lib/python3.9/site-packages/gpytorch/lazy/lazy_tensor.py:1243: in inv_quad_logdet
    cholesky = CholLazyTensor(TriangularLazyTensor(self.cholesky()))
../../_tool/Python/3.9.18/x64/lib/python3.9/site-packages/gpytorch/lazy/lazy_tensor.py:962: in cholesky
    chol = self._cholesky(upper=False)
../../_tool/Python/3.9.18/x64/lib/python3.9/site-packages/gpytorch/utils/memoize.py:59: in g
    return _add_to_cache(self, cache_name, method(self, *args, **kwargs), *args, kwargs_pkl=kwargs_pkl)
../../_tool/Python/3.9.18/x64/lib/python3.9/site-packages/gpytorch/lazy/lazy_tensor.py:425: in _cholesky
    cholesky = psd_safe_cholesky(evaluated_mat, upper=upper).contiguous()
../../_tool/Python/3.9.18/x64/lib/python3.9/site-packages/gpytorch/utils/cholesky.py:108: in psd_safe_cholesky
    L = _psd_safe_cholesky(A, out=out, jitter=jitter, max_tries=max_tries)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

A = tensor([[nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,...nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
         nan, nan, nan, nan, nan, nan]], grad_fn=<AddBackward0>)
out = None, jitter = None, max_tries = 3

    def _psd_safe_cholesky(A, out=None, jitter=None, max_tries=3):
        # Maybe log
        if settings.verbose_linalg.on():
            settings.verbose_linalg.logger.debug(f"Running Cholesky on a matrix of size {A.shape}.")
    
        if out is not None:
            out = (out, torch.empty(A.shape[:-2], dtype=torch.int32))
    
        L, info = torch.linalg.cholesky_ex(A, out=out)
        if not torch.any(info):
            return L
    
        isnan = torch.isnan(A)
        if isnan.any():
>           raise NanError(
                f"cholesky_cpu: {isnan.sum().item()} of {A.numel()} elements of the {A.shape} tensor are NaN."
            )
E           gpytorch.utils.errors.NanError: cholesky_cpu: 900 of 900 elements of the torch.Size([30, 30]) tensor are NaN.

../../_tool/Python/3.9.18/x64/lib/python3.9/site-packages/gpytorch/utils/cholesky.py:30: NanError

During handling of the above exception, another exception occurred:

self = <tests.units.warps.test_cwgp_training.ConstraintTests testMethod=test_fitting_with_constrained_warp>

    def test_fitting_with_constrained_warp(self):
        """Should NOT throw a RuntimeError."""
        box_cox = warpfunctions.BoxCoxWarpFunction(lambda_=0)
        affine = warpfunctions.PositiveAffineWarpFunction(a=1, b=2)
    
        @SetWarp(box_cox @ affine, ignore_methods=("__init__",))
        class TestController(GaussianGPController):
            """A test controller."""
            pass
    
        scaler = StandardScaler()
        gp = TestController(scaler.fit_transform(self.DATASET.train_x), self.DATASET.train_y, ScaledRBFKernel,
                            y_std=self.DATASET.train_y_std)
    
        try:
            gp.fit(100)
        except NanError as error:
>           self.fail(f"Should not have thrown: {error!s}")
E           AssertionError: Should not have thrown: cholesky_cpu: 900 of 900 elements of the torch.Size([30, 30]) tensor are NaN.

tests/units/warps/test_cwgp_training.py:299: AssertionError

--- a/tests/units/warps/test_cwgp_training.py
+++ b/tests/units/warps/test_cwgp_training.py
@@ -11,7 +11,7 @@ from vanguard.kernels import ScaledRBFKernel
 from vanguard.vanilla import GaussianGPController
 from vanguard.warps import SetWarp, WarpFunction, warpfunctions
 
-from ...cases import VanguardTestCase
+from ...cases import VanguardTestCase, flaky
 
 
 class CompositionTests(VanguardTestCase):
@@ -279,6 +279,7 @@ class ConstraintTests(VanguardTestCase):
         with self.assertRaisesRegex(NanError, expected_regex):
             gp.fit(100)
 
+    @flaky
     def test_fitting_with_constrained_warp(self) -> None:
         """Should NOT throw a RuntimeError."""
         box_cox = warpfunctions.BoxCoxWarpFunction(lambda_=0)

